server/src/main/java/org/elasticsearch/common/util/concurrent/AbstractAsyncTask.java

```diff
@@ -91,7 +91,7 @@ public synchronized void rescheduleIfNecessary() {
             if (logger.isTraceEnabled()) {
                 logger.trace("scheduling {} every {}", toString(), interval);
             }
-            cancellable = threadPool.schedule(this, interval, getThreadPool());
+            cancellable = threadPool.schedule(threadPool.preserveContext(this), interval, getThreadPool());
```

Comment:
`ThreadPool#schedule` does not itself preserve the context of the caller and instead runs the scheduled task in the default context which is not a system context.

Reply:
Yeah, I knew about that. The question was more about the motivation for moving from simply setting the system context inside the task to preserving it here. I did some double checking on the callers and AFAICS it looks ok, just seemed odd to me to make this change for this specific PR. On second thought, I think it makes sense to preserve the context here, given that the `AbstractAsyncTask` is not specific to `IndexService`, but could be good to maybe add an assertion about being in system-context to `IndexService`?

Reply:
I see. I prefer preserving an existing context here since the security implications are much clearer - security bugs lurk in areas where privileges change, so the less of that we do the better. IMO it's a bit trappy that `EsThreadPoolExecutor#execute` preserves the caller's context but `ThreadPool#schedule` does not, although this is one of the very few places where that matters right now.

It's possible we could assert that we are in system context here, but that seems an overly strong statement to make. We already have tests to assert that we're in a context with appropriate permissions which I think is enough.